scipy/signal/_peak_finding.py

```diff
@@ -233,6 +241,578 @@ def argrelextrema(data, comparator, axis=0, order=1, mode='clip'):
+        See `infodict` for a full description.
+
+    Returns
+    -------
+    peaks : ndarray
+        Indices of peaks in `vector` that fullfill all requirements.
+    infodict : dict, only returned if full_output is true
+        A dictionary containing intermediate results from peak filtering:
+
+        ==================== ====================================================
+        Key                  Description
+        ==================== ====================================================
+        'heights'            Value of `vector` and each peak. Returned if
+                             `height` was given.
+        'stacked_thresholds' Array of shape (2, peak size) containing the
+                             thresholds of each peak on both sides. The first
+                             row contains the thresholds to the left and the
+                             second row to the right of each peak. Returned if
+                             `threshold` was given.
+        'prominences'        Prominence of each returned peak. Returned if
+                             `prominence` was given.
+        'widths'             Width of each returned peak. Returned if `width` was
+                             given.
+        'wheights'           Value of `vector` at which each width was evaluated.
+                             Returned if `width` was given.
+        'left_width_pos'     Left border position used to calculate each peak
+                             width. Returned if `width` was given.
+        'right_width_pos'    Same as above for right position.
+        ==================== ====================================================
+
+    See Also
+    --------
+    find_peaks_cwt
+        Find peaks using wavelet transformation.
+    peak_prominences
+        Directly calculate prominences of peaks.
+    peak_widths
+        Directly calculate widths of peaks.
+
+    Notes
+    -----
+
+    .. versionadded:: 1.1.0
+
+    Examples
+    --------
+    Create test signal
+
+    >>> time = np.linspace(0, 8 * np.pi, 1000)
+    >>> vector = np.sin(time) + 0.2 * np.sin(10 * time) + 0.1 * time
+
+    Find peaks and require a minimal peak prominence of 0.5
+
+    >>> from scipy import signal
+    >>> peaks, info = signal.find_peaks(vector, prominence=0.5, full_output=True)
+    >>> peaks
+    array([ 57, 306, 556, 806])
+
+    Plot results
+
+    >>> import matplotlib.pyplot as plt
+    >>> countour_height = vector[peaks] - info['prominences']
+    >>> plt.plot(time, vector)
+    >>> plt.plot(time[peaks], vector[peaks], "x")
+    >>> plt.vlines(time[peaks], ymin=countour_height, ymax=vector[peaks])
+    """
+    vector = np.asarray(vector)
+    peaks = argrelmax(vector)[0]
+    infodict = {}
+
+    # Filter by height
+    if height is not None:
+        hmin, hmax = _parse_filter_args(height, vector, peaks)
+        heights = vector[peaks]
+        keep = np.ones(peaks.size, dtype=bool)
```

Comment:
Is the initialization of `keep` needed here? I'm just thinking that the first comparison below would automatically generate a bool or mask array anyway.

Reply:
Yes it is needed. Look at the possible valid inputs for the argument `height`:

1. `height=hmin` -> `hmax` is `None`, only first comparison is evaluated
2. `height=(None, hmax)` -> `hmin` is `None`, only second comparison is evaluated
3. `height=(hmin, hmax)` -> both comparisons are evaluated
4. `height=(None, None)` -> no comparison is evaluated (has no effect but is valid)

Case 3 dictates that the second evaluation must use `&=` (`keep` must be initialized). Considering the 2nd case one can easily see that initialization is not guarantied to happen in the first comparison.

Reply:
Alright, looks like a lot of input permutation handling is indeed needed.